File: alpha-factor-forge/src/core/validation/split.ts
// Phase B validation foundation: deterministic, time-ordered bar ranges.
// Pure TypeScript only — no React, DOM, IO, persistence, or ranking behavior.

export const V1_VALIDATION_RATIOS = {
  train: 0.6,
  validation: 0.2,
  test: 0.2,
} as const;

export interface InclusiveBarRange {
  /** Zero-based inclusive first bar, compatible with BacktestConfig.from. */
  readonly from: number;
  /** Zero-based inclusive last bar, compatible with BacktestConfig.to. */
  readonly to: number;
  readonly count: number;
}

export interface ValidationSplitPlan {
  readonly totalBars: number;
  /** Bars available to Train/Validation/Test after both embargo gaps. */
  readonly usableBars: number;
  /** Size of each boundary gap; the caller owns how this is derived. */
  readonly embargoBars: number;
  readonly train: InclusiveBarRange;
  readonly trainValidationEmbargo: InclusiveBarRange | null;
  readonly validation: InclusiveBarRange;
  readonly validationTestEmbargo: InclusiveBarRange | null;
  /** Always the final evaluated range. It must not drive v1 ranking or prompts. */
  readonly test: InclusiveBarRange;
}

interface Allocation {
  train: number;
  validation: number;
  test: number;
}

const MIN_USABLE_BARS = 5;
const V1_RATIO_WEIGHTS = [3, 1, 1] as const;
const V1_RATIO_DENOMINATOR = 5;

function assertNonNegativeSafeInteger(value: number, name: string): void {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new RangeError(`${name} must be a non-negative safe integer`);
  }
}

/**
 * Allocate the usable bars with the largest-remainder method. Equal
 * fractional remainders are resolved in Train -> Validation -> Test order.
 */
function allocateUsableBars(usableBars: number): Allocation {
  // Split first so we never multiply a large safe integer by a floating-point
  // ratio (or by 3 directly). Every intermediate remains an exact safe integer.
  const quotient = Math.floor(usableBars / V1_RATIO_DENOMINATOR);
  const tail = usableBars % V1_RATIO_DENOMINATOR;
  const counts = V1_RATIO_WEIGHTS.map(
    (weight) => weight * quotient + Math.floor((weight * tail) / V1_RATIO_DENOMINATOR),
  );
  const remainderOrder = V1_RATIO_WEIGHTS
    .map((weight, index) => ({
      index,
      numerator: (weight * tail) % V1_RATIO_DENOMINATOR,
    }))
    .sort((a, b) => b.numerator - a.numerator || a.index - b.index);

  let remaining = usableBars - counts.reduce((sum, count) => sum + count, 0);
  for (let i = 0; remaining > 0; i++, remaining--) {
    counts[remainderOrder[i].index] += 1;
  }

  const [train, validation, test] = counts;
  if (train < 1 || validation < 1 || test < 1) {
    throw new RangeError(
      `at least ${MIN_USABLE_BARS} usable bars are required for non-empty 60/20/20 segments`,
    );
  }
  return { train, validation, test };
}

function rangeFrom(cursor: number, count: number): InclusiveBarRange {
  return { from: cursor, to: cursor + count - 1, count };
}

/**
 * Plan the v1 Train/Validation/Test split over bars sorted oldest to newest.
 *
 * Contract:
 * 1. Exclude two equal embargo gaps from the total bar count.
 * 2. Allocate the remaining bars 60/20/20 with deterministic largest-remainder
 *    rounding (ties favor Train, then Validation, then Test).
 * 3. Return inclusive, ascending ranges that account for every input bar once.
 *
 * Invalid or insufficient input throws instead of silently weakening an
 * embargo or returning an empty evaluation segment.
 */
export function planValidationSplit(totalBars: number, embargoBars: number): ValidationSplitPlan {
  assertNonNegativeSafeInteger(totalBars, 'totalBars');
  assertNonNegativeSafeInteger(embargoBars, 'embargoBars');

  const usableBars = totalBars - embargoBars * 2;
  if (usableBars < MIN_USABLE_BARS) {
    throw new RangeError(
      `at least ${MIN_USABLE_BARS} usable bars are required after two embargo gaps`,
    );
  }

  const allocation = allocateUsableBars(usableBars);
  let cursor = 0;

  const train = rangeFrom(cursor, allocation.train);
  cursor = train.to + 1;

  const trainValidationEmbargo = embargoBars > 0 ? rangeFrom(cursor, embargoBars) : null;
  cursor += embargoBars;

  const validation = rangeFrom(cursor, allocation.validation);
  cursor = validation.to + 1;

  const validationTestEmbargo = embargoBars > 0 ? rangeFrom(cursor, embargoBars) : null;
  cursor += embargoBars;

  const test = rangeFrom(cursor, allocation.test);

  return {
    totalBars,
    usableBars,
    embargoBars,
    train,
    trainValidationEmbargo,
    validation,
    validationTestEmbargo,
    test,
  };
}

File: alpha-factor-forge/src-tauri/src/discovery_core/split.rs
//! `validation-split-v1`: pure Rust parity port of `src/core/validation/split`.
//!
//! Deterministic 60/20/20 largest-remainder allocation after two equal
//! embargo gaps, with exact integer arithmetic and the same fail-closed
//! messages as the TypeScript reference (VAL-001 contract).

use serde::{Deserialize, Serialize};

pub const SPLIT_CONTRACT_VERSION: &str = "validation-split-v1";

const MIN_USABLE_BARS: i64 = 5;
const JS_MAX_SAFE_INTEGER: i64 = 9_007_199_254_740_991;
const RATIO_WEIGHTS: [i64; 3] = [3, 1, 1];
const RATIO_DENOMINATOR: i64 = 5;

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct InclusiveBarRange {
    pub from: i64,
    pub to: i64,
    pub count: i64,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ValidationSplitPlan {
    pub total_bars: i64,
    pub usable_bars: i64,
    pub embargo_bars: i64,
    pub train: InclusiveBarRange,
    pub train_validation_embargo: Option<InclusiveBarRange>,
    pub validation: InclusiveBarRange,
    pub validation_test_embargo: Option<InclusiveBarRange>,
    pub test: InclusiveBarRange,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SplitError(pub String);

impl std::fmt::Display for SplitError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(&self.0)
    }
}

impl std::error::Error for SplitError {}

fn assert_non_negative_safe_integer(value: i64, name: &str) -> Result<(), SplitError> {
    if !(0..=JS_MAX_SAFE_INTEGER).contains(&value) {
        return Err(SplitError(format!(
            "{name} must be a non-negative safe integer"
        )));
    }
    Ok(())
}

struct Allocation {
    train: i64,
    validation: i64,
    test: i64,
}

/// Largest-remainder allocation with Train -> Validation -> Test tie order.
fn allocate_usable_bars(usable_bars: i64) -> Result<Allocation, SplitError> {
    let quotient = usable_bars / RATIO_DENOMINATOR;
    let tail = usable_bars % RATIO_DENOMINATOR;
    let mut counts: Vec<i64> = RATIO_WEIGHTS
        .iter()
        .map(|weight| weight * quotient + (weight * tail) / RATIO_DENOMINATOR)
        .collect();

    let mut remainder_order: Vec<(usize, i64)> = RATIO_WEIGHTS
        .iter()
        .enumerate()
        .map(|(index, weight)| (index, (weight * tail) % RATIO_DENOMINATOR))
        .collect();
    remainder_order.sort_by(|left, right| right.1.cmp(&left.1).then(left.0.cmp(&right.0)));

    let mut remaining = usable_bars - counts.iter().sum::<i64>();
    let mut order = remainder_order.iter();
    while remaining > 0 {
        let (index, _) = order.next().expect("at most two remainder bars exist");
        counts[*index] += 1;
        remaining -= 1;
    }

    let (train, validation, test) = (counts[0], counts[1], counts[2]);
    if train < 1 || validation < 1 || test < 1 {
        return Err(SplitError(format!(
            "at least {MIN_USABLE_BARS} usable bars are required for non-empty 60/20/20 segments"
        )));
    }
    Ok(Allocation {
        train,
        validation,
        test,
    })
}

fn range_from(cursor: i64, count: i64) -> InclusiveBarRange {
    InclusiveBarRange {
        from: cursor,
        to: cursor + count - 1,
        count,
    }
}

/// Plan the v1 Train/Validation/Test split over bars sorted oldest to newest.
pub fn plan_validation_split(
    total_bars: i64,
    embargo_bars: i64,
) -> Result<ValidationSplitPlan, SplitError> {
    assert_non_negative_safe_integer(total_bars, "totalBars")?;
    assert_non_negative_safe_integer(embargo_bars, "embargoBars")?;

    let usable_bars = total_bars - embargo_bars * 2;
    if usable_bars < MIN_USABLE_BARS {
        return Err(SplitError(format!(
            "at least {MIN_USABLE_BARS} usable bars are required after two embargo gaps"
        )));
    }

    let allocation = allocate_usable_bars(usable_bars)?;
    let mut cursor = 0i64;

    let train = range_from(cursor, allocation.train);
    cursor = train.to + 1;

    let train_validation_embargo = if embargo_bars > 0 {
        Some(range_from(cursor, embargo_bars))
    } else {
        None
    };
    cursor += embargo_bars;

    let validation = range_from(cursor, allocation.validation);
    cursor = validation.to + 1;

    let validation_test_embargo = if embargo_bars > 0 {
        Some(range_from(cursor, embargo_bars))
    } else {
        None
    };
    cursor += embargo_bars;

    let test = range_from(cursor, allocation.test);

    Ok(ValidationSplitPlan {
        total_bars,
        usable_bars,
        embargo_bars,
        train,
        train_validation_embargo,
        validation,
        validation_test_embargo,
        test,
    })
}
